### ISSUE TITLE
Why EventCallback instead of normal c# events?

### ISSUE THREAD

**Issue by arivoir** [aspnet/AspNetCore#11035]
I'm trying to expand a suite of cross-platform controls to Blazor. I was working on the properties and achieved a good degree of reutilization, but I'm having trouble to share events between platforms, because Blazor doesn't have any support for c# events, instead it uses EventCallback. I think normal events would work just fine, I don't see any reason to force the usage of EventCallback. Can you explain why?

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us, @arivoir.
The reason is that with events developers would need to own logic to unregister their events when necessary, which is relatively hard problem, and people usually just don't do that. That would result in resources not being reclaimed properly and server memory would grow.
`EventCallback` solves that problem and we do the needful behind the scenes.

**Comment by mkArtakMSFT** [comment]
@rynowak is there anything else to add here?

**Comment by arivoir** [comment]
> Thanks for contacting us, @arivoir.
> The reason is that with events developers would need to own logic to unregister their events when necessary, which is relatively hard problem, and people usually just don't do that. That would result in resources not being reclaimed properly and server memory would grow.
> EventCallback solves that problem and we do the needful behind the scenes.

I see your point. I need to think more on this. Thanks for you reply.

**Comment by arivoir** [comment]
Let me give details a scenario I have in mind.

I have a grid control whose columns are generated automatically from the source, which is set through a property named "ItemsSource"

`<Grid ItemsSource="@someItemsSource"/>`

I want customers to have an event AutoGeneratingColumns, which will be triggered for each columns, and where customers will be able to customize every column.

If I think in the way I would expect the razor file to be specified, this comes to my mind

```
<Grid ItemsSource="@someItemsSource" AutoGeneratingColumns="@OnAutoGeneratingColumns"/>

@functions {
    void OnAutoGeneratingColumns(eventdata)
   {
        //Here the customer will customize the column
   }
}
```

But this model brings too problems. 

1. The invocation to EventCallback is asynchronous, therefore is not suitable for the scenario.
2. The EventCallback automatically calls StateHasChanged, which is not reasonable for this scenario. Imagine when 100 columns are created this will caused the state of the component to be recreated 100 times.

It looks the EventCallback is not thought for this kind of scenarios, so I'm wondering how to implement normal events in Blazor?

**Comment by rynowak** [comment]
@arivoir 

> I want customers to have an event AutoGeneratingColumns, which will be triggered for each columns, and where customers will be able to customize every column.

This doesn't really sound like an event, it's a template. Specifically I mean that it's not a a change of state, it's just some configuration you want to pass in. If the output of this callback is markup, then that sounds like a job for `RenderFragment<>`

**Comment by arivoir** [comment]
Hi @rynowak ,

The generation of columns affect the markup, but it is not directly related, and there is no a specific peace of markup tied to a column. There are the cells of the columns, but they are generated according to the visible viewport and change as the user scrolls.

In all the other platforms this is just a simple c# event that can be handled by the customers to alter the way the auto-generation of columns work.

The point is Blazor seems not supporting normal events, there is the EventCallback but I think it doesn't cover all the scenarios. How do you think this kind of scenario should be resolved in Blazor?

I appreciate any feedback.

**Comment by rynowak** [comment]
> The point is Blazor seems not supporting normal events, there is the EventCallback but I think it doesn't cover all the scenarios. How do you think this kind of scenario should be resolved in Blazor?

Sorry, I'm still having trouble understanding. You can define .NET events. You can use delegate types as parameters. How this typically works is you would define a parameter like:

```razor
[Parameter] RenderFragment<ColumnInfo> ColumnTemplate { get; set; }
```

**Comment by arivoir** [comment]
I was referring to a C# event https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/events/

Something like this

`public event EventHandler<GridAutoGeneratingColumnEventArgs> AutoGeneratingColumn;`


This event doesn't need a fragment, and it shouldn't emit any markup.

Do you see my point?

**Comment by arivoir** [comment]
I just made an experiment. 

Converted the event

`public event EventHandler<GridAutoGeneratingColumnEventArgs> AutoGeneratingColumn;`

Into this

```
[Parameter]
public EventHandler<GridAutoGeneratingColumnEventArgs> AutoGeneratingColumn { get; set; }
```

And it looks working (Provided the event handler is set before the source)

```
<FlexGrid AutoGeneratingColumn="@OnAutoGeneratingColumn" ItemsSource="@customers"/>

@functions {
    ObservableCollection<Customer> customers;

    protected override void OnInit()
    {
        customers = Customer.GetCustomerList(100);
    }

    private void OnAutoGeneratingColumn(object sender, GridAutoGeneratingColumnEventArgs e)
    {

    }
}

```

My point is why not to allow events to be decorated with [Parameter] and allow specify them in razor files?

![image](https://user-images.githubusercontent.com/17496741/59446538-17027480-8dd8-11e9-84f1-6c63ca0767c1.png)

**Comment by rynowak** [comment]
> My point is why not to allow events to be decorated with [Parameter] and allow specify them in razor files?

TLDR, this wouldn't add value because whether its an event or a delegate it will behave the same way.

Because your use case is the same as a normal delegate. The `EventHandler<>` delegate type isn't fundamentally different from using `Action<object, GridAutoGeneratingColumnEventArgs>`. Invoking a .NET event is the same as invoking a delegate.

The semantic difference between and event and a delegate is in how you *subscribe* to an event. You subscribe and unsubscribe to events using `+=` and `-=`, and events allow you to attach multiple event handlers in a chain.

Nothing in that last paragraph is true about Blazor. You pass parameters explicitly (as attributes), and we don't support multiple values for the same parameters. So if we had a feature like this, it would be the same for consumers and the same for implementors.

I generally prefer not to provide a choice when the choice is not meaningful or useful. If we supported this, I wouldn't be able to write documentation that explains when to use one or the other - therefore I don't want to have it.
